testsuite/tests/typing-layouts/or_null.ml

```diff
@@ -0,0 +1,326 @@
+   ** expect
+*)
+(* CR layouts (v3): enable this test *)
+
+module type Or_null = sig
+  (* CR layouts (v3): Not sure how to express that None and Some should
+     be part of this module. They're not quite constructors. So the syntax
+     here might plausibly change. *)
+  type 'a t = 'a or_null =
+    | None
+    | Some of 'a
+
+  val none : 'a or_null
+  val some : 'a -> 'a or_null
+  val value : 'a or_null -> default:'a -> 'a
+  val get : 'a or_null -> 'a
+  val bind : 'a or_null -> ('a -> 'b or_null) -> 'b or_null
+  (* unlike [option] we cannot have [join] *)
+  val map : ('a -> 'b) -> 'a or_null -> 'b or_null
+  val fold : none:'a -> some:('b -> 'a) -> 'b or_null -> 'a
+  val iter : ('a -> unit) -> 'a or_null -> unit
+
+  val is_none : 'a or_null -> bool
+  val is_some : 'a or_null -> bool
+  val equal : ('a -> 'a -> bool) -> 'a or_null -> 'a or_null -> bool
+  val compare : ('a -> 'a -> int) -> 'a or_null -> 'a or_null -> int
+
+  val to_result : none:'e -> 'a or_null -> ('a, 'e) result
+  val to_list : 'a or_null -> 'a list
+  val to_seq : 'a or_null -> 'a Seq.t
+
+  val to_option : 'a or_null -> 'a option
+  val of_option : 'a option -> 'a or_null
+end
+
+module Or_null : Or_null = Or_null
+
+[%%expect {|
+success
+|}]
+
+(* ensure that immediacy "looks through" or_null *)
+type t = (int or_null : immediate)
+type t = (bool or_null : immediate)
+
+[%%expect {|
+success
+|}]
+
+type t = (string or_null : immediate)
+
+[%%expect {|
+error
+|}]
+
+type t = (string or_null : value)
+
+[%%expect {|
+success
+|}]
+
+(* ensure that or_null can't be repeated *)
+type 'a t = 'a or_null or_null
+
+[%%expect {|
+error
+|}]
+
+(* check inference around or_null *)
+type 'a t = 'a or_null
+type ('a : immediate) t = 'a or_null
+
+[%%expect {|
+success
+success (inferring an immediate layout for [t] and non_null_immediate for ['a])
+|}]
+
+(* more layout checking *)
+type t = (string or_null : non_null_value)
+
+[%%expect {|
+error
+|}]
+
+type t = (string : non_null_value)
+type t = (int : non_null_value)
+type t = (int : non_null_immediate)
+type t = (int or_null : value)
+
+[%%expect {|
+success
+|}]
+
+(* magic looking-through of [or_null] can't be abstracted over *)
+type 'a t = 'a or_null
+type q = (string t : value)
+type q = (int t : immediate)  (* but t isn't abstract, so this is OK *)
+
+[%%expect {|
+success
+|}]
+
+type q = string t t
+
+[%%expect {|
+error
+|}]
+
+type q = int t t
+
+[%%expect {|
+error
+|}]
+
+type 'a q = 'a t
+type ('a : immediate) q = 'a t
+
+[%%expect {|
+success
+|}]
+
+module type T = sig
+  type t
+end
+
+[%%expect {|
+success
+|}]
+
+(* this should be rejected, because the default for [t] is [non_null_value] *)
+module M : T = struct
+  type t = string or_null
+end
+
+[%%expect {|
+error
+|}]
+
+module M : T = struct
+  type t = int or_null
+end
+
+[%%expect {|
+error
+|}]
+
+module M : sig
+  type 'a t
+end = struct
+  type 'a t = 'a or_null
+end
+
+[%%expect {|
+error
+|}]
```

Comment:
Maybe a note here that the error message had better be very good (perhaps with an explicit hint) because needing a `value` annotation is unusual.